Linker error

So, now all modules have successfully compiled and there are no issues there! Now I encountered a new bug, allow me to dump what it says on the terminal (user is mega):
Linking hipreme_engine
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_4d_657.o): in function `_D3hip9windowing9platforms3x1116initializeOpenGLFNbNiiiZb':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:60: undefined reference to `glXChooseFBConfig'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:64: undefined reference to `XCloseDisplay'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:71: undefined reference to `glXGetVisualFromFBConfig'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:74: undefined reference to `glXGetFBConfigAttrib'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:75: undefined reference to `glXGetFBConfigAttrib'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:85: undefined reference to `XFree'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:91: undefined reference to `XFree'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:92: undefined reference to `glXGetVisualFromFBConfig'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:96: undefined reference to `XCloseDisplay'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:106: undefined reference to `XCreateColormap'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:126: undefined reference to `XCreateWindow'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:134: undefined reference to `XInternAtom'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:146: undefined reference to `XSetWMProtocols'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:158: undefined reference to `glXQueryExtensionsString'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:161: undefined reference to `glXGetProcAddressARB'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:169: undefined reference to `glXCreateNewContext'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:189: undefined reference to `XSync'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:191: undefined reference to `glXIsDirect'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:193: undefined reference to `glXMakeCurrent'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:199: undefined reference to `XClearWindow'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:200: undefined reference to `XMapRaised'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:201: undefined reference to `XStoreName'
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_4f_415.o): in function `_D3hip9windowing9platforms3x1110swapBufferFZv':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:212: undefined reference to `glXSwapBuffers'
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_50_5bb.o): in function `_D3hip9windowing9platforms3x1114setVsyncActiveFNbNibZv':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:220: undefined reference to `glXGetProcAddressARB'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:221: undefined reference to `glXGetProcAddressARB'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:222: undefined reference to `glXGetProcAddressARB'
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_53_1b7.o): in function `_D3hip9windowing9platforms3x114pollFZv':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:249: undefined reference to `XPending'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:251: undefined reference to `XNextEvent'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:257: undefined reference to `XGetWindowAttributes'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:327: undefined reference to `XKeycodeToKeysym'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:332: undefined reference to `XKeycodeToKeysym'
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_56_740.o): in function `_D3hip9windowing9platforms3x1118destroy_GL_ContextFZb':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:359: undefined reference to `XDestroyWindow'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:360: undefined reference to `XCloseDisplay'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:361: undefined reference to `XFree'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:362: undefined reference to `XFreeColormap'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:363: undefined reference to `glXDestroyContext'
/usr/bin/ld: modules/windowing/.dub/build/default-debug-linux.posix-x86_64-dmd_v2.101.0-1487C0C4D73EE1149313B5787F50E73F491D7E43EBD1E7409DB3001BA1124E01/libwindowing.a(x11_57_42a.o): in function `_D3hip9windowing9platforms3x1110openWindowFiiZi':
/home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:372: undefined reference to `XOpenDisplay'
/usr/bin/ld: /home/mega/HipremeEngine/modules/windowing/source/hip/windowing/platforms/x11.d:379: undefined reference to `glXQueryVersion'
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
Error /usr/bin/dmd failed with exit code 1.

Not sure what is going on here admittedly, though so close to being compiled now! (not sure if this is something related to my linker, or the codebase, or something else entirely even)

--- a/dub.json
+++ b/dub.json
@@ -104,10 +104,6 @@
 				"NO_GLES",
 				"NO_SLES"
 			],
-			"libs-linux" : [
-				"X11",
-				"gl"
-			],
 			"subConfigurations": {
 				"audio": "nodx",
 				"renderer": "nodx"
@@ -141,10 +137,6 @@
 			"lflags-linux": [
 				"--export-dynamic"
 			],
-			"libs-linux" : [
-				"X11",
-				"gl"
-			],
 			"buildOptions": [
 				"debugMode",
 				"debugInfo",
@@ -170,10 +162,6 @@
 				"Standalone",
 				"Desktop"
 			],
-			"libs-linux" : [
-				"X11",
-				"gl"
-			],
 			"versions-windows": [
 				"Direct3D_11",
 				"XAudio2",
@@ -197,10 +185,6 @@
 				"OpenAL",
 				"Desktop",
 				"HipConcurrency"
-			],
-			"libs-linux" : [
-				"X11",
-				"gl"
 			],
 			"versions-windows": [
 				"Direct3D_11",

--- a/modules/console/source/hip/console/log.d
+++ b/modules/console/source/hip/console/log.d
@@ -56,6 +56,7 @@ ulong line = __LINE__)
 
 
 void rawlog(Args... )(Args a){Console.DEFAULT.log(a);}
+void rawwarn(Args... )(Args a){Console.DEFAULT.warn(a);}
 void rawinfo(Args... )(Args a){Console.DEFAULT.info(a);}
 void rawerror(Args... )(Args a){Console.DEFAULT.error(a);}
 void rawfatal(Args... )(Args a){Console.DEFAULT.fatal(a);}

--- a/modules/error/source/hip/error/handler.d
+++ b/modules/error/source/hip/error/handler.d
@@ -11,10 +11,6 @@ Distributed under the CC BY-4.0 License.
 module hip.error.handler;
 import hip.console.log;
 import hip.util.conv;
-version(Android)
-{
-    import hip.jni.helper.androidlog;
-}
 
 /** 
  * Base clas for documenting errors
@@ -125,16 +121,8 @@ public static class ErrorHandler
     }
     public static void showWarningMessage(string warningTitle, string warningMessage)
     {
-        version(Android)
-        {
-            alogw("HipremeEngine", "\nWarning: " ~ warningTitle);
-            alogw("HipremeEngine", warningMessage);
-        }
-        else
-        {
-            rawlog("\nWarning: " ~ warningTitle);
-            rawlog(warningMessage);
-        }
+        rawwarn("\nWarning: " ~ warningTitle);
+        rawwarn(warningMessage);
         warnHistory~= warningTitle~": "~warningMessage;
     }
 

--- a/modules/filesystem/source/hip/filesystem/hipfs.d
+++ b/modules/filesystem/source/hip/filesystem/hipfs.d
@@ -141,7 +141,9 @@ class HipFileSystem
     {
         import hip.util.path:joinPath;
         import hip.util.system : sanitizePath;
-        return joinPath(combinedPath, path.sanitizePath);
+        if(combinedPath)
+            return joinPath(combinedPath, path.sanitizePath);
+        return path.sanitizePath;
     }
     @ExportD public static bool isPathValid(string path){return validatePath(initialPath, defPath~path);}
     @ExportD public static bool isPathValidExtra(string path)
@@ -166,8 +168,13 @@ class HipFileSystem
     {
         import hip.util.path:joinPath;
         import hip.util.system : sanitizePath;
-        defPath = path.sanitizePath;
-        combinedPath = joinPath(initialPath, defPath);
+        if(path)
+        {
+            defPath = path.sanitizePath;
+            combinedPath = joinPath(initialPath, defPath);
+        }
+        else
+            combinedPath = initialPath;
         return validatePath(initialPath, combinedPath);
     }
 

--- a/modules/renderer/dub.json
+++ b/modules/renderer/dub.json
@@ -84,7 +84,8 @@
 				"math": {"path": "../math"}
 			},
 			"subConfigurations": {
-				"bind": "android"
+				"bind": "android",
+				"windowing": "android"
 			},
 			"versions": [
 				"HipGLUseVertexArray"

--- a/modules/tween/source/hip/tween.d
+++ b/modules/tween/source/hip/tween.d
@@ -12,6 +12,7 @@ private enum c5 = (2 * PI) / 4.5f;
 private enum n1 = 7.5625f;
 private enum d1 = 2.75f;
 
+//TODO: Change (x) to (float x) for not generating templates needlessly
 /**
 *   Credits to https://easings.net as I don't understand most of those functions
 */
@@ -57,7 +58,7 @@ enum HipEasing : float function(float x)
                              ? -(pow(2, 20 * x - 10) * sin((20 * x - 11.125f) * c5))/2
                              : (pow(2, -20 * x + 10) * sin((20 * x - 11.125f) * c5))/2 + 1,
     easeInBounce    = (x) => 1 - HipEasing.easeOutBounce(1 - x),
-    easeOutBounce   = function float(float x)
+    easeOutBounce   = function float(x)
     {
         if(x < 1.0f / d1)
             return n1 * x * x;

--- a/modules/util/source/hip/util/path.d
+++ b/modules/util/source/hip/util/path.d
@@ -30,7 +30,7 @@ string[] pathSplitter(string path) @safe pure nothrow
     {
         if(path[i] == '/' || path[i] == '\\')
         {
-            ret~= current;
+            ret~= current~path[i];
             current = null;
             continue;
         }
@@ -296,10 +296,7 @@ string joinPath(char separator, in string[] paths ...) @safe pure nothrow
         string next = i+1 < paths.length ? paths[i+1] : "";
 
         if(filePath == "")
-        {
-            output~= separator;
             continue;
-        }
         
         output~=paths[i];
         if(next != "" && next[0] != separator  &&

--- a/modules/util/source/hip/util/string.d
+++ b/modules/util/source/hip/util/string.d
@@ -24,7 +24,6 @@ struct String
     private size_t _capacity;
     private int* countPtr;
 
-    import std.stdio;
     this(this)
     {
         if(countPtr !is null)

--- a/modules/util/source/hip/util/to_string_range.d
+++ b/modules/util/source/hip/util/to_string_range.d
@@ -1,9 +1,19 @@
 module hip.util.to_string_range;
 import std.range.interfaces;
 import std.range.primitives;
-import std.range:put;
 import std.traits:isArray;
 import std.typecons:isTuple;
+
+private void put(Sink, E)(ref Sink sink, E e)
+{
+    static if(is(E == U[], U))
+        foreach(element; e)
+            sink.put(element);
+    else
+        sink.put(e);
+}
+
+
 
 void toStringRange(Sink)(auto ref Sink sink, float f)
 {

--- a/modules/windowing/dub.json
+++ b/modules/windowing/dub.json
@@ -4,7 +4,14 @@
 	],
 	"configurations": [
 		{
-			"name": "default"
+			"name": "default",
+			"libs-linux" : [
+				"X11",
+				"gl"
+			]
+		},
+		{
+			"name": "android"
 		},
 		{
 			"name": "source",

--- a/source/app.d
+++ b/source/app.d
@@ -44,7 +44,7 @@ import hip.config.opts;
 *	If you wish to use LLD on Windows, you will need to install Windows 10 SDK.
 *	And include on dub the following lflags:
 *	"lflags": [
-		"-libpath:C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\lib\\x64"
+		"-libpath:C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\lib\\x64\\"
 	],
 *
 *	The libpath is required for getting to link a "compatible" libcmt, which unfortunately is not really compatible for some reason.

